The job **`aarch64-gnu-llvm-21-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/26319996904/job/77487057685) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/77487057685) [(plain)](https://github.com/rust-lang/rust/commit/aed2feb23bbd6cc44f4dc39560902ce9c0ef495d/checks/77487057685/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
+    |                           ^^^^^^^^^^----------
+    |                                     |
+    |                                     help: remove the `use<...>` syntax
22 
23 error: all possible in-scope parameters are already captured, so `use<...>` syntax is redundant
24   --> $DIR/redundant.rs:15:22

25    |
26 LL |     fn in_trait() -> impl Sized + use<'a, Self>;
-    |                      ^^^^^^^^^^^^^-------------
-    |                                   |
-    |                                   help: remove the `use<...>` syntax
+    |                      ^^^^^^^^^^----------------
+    |                                |
+    |                                help: remove the `use<...>` syntax
30 
31 error: all possible in-scope parameters are already captured, so `use<...>` syntax is redundant
32   --> $DIR/redundant.rs:19:22

33    |
34 LL |     fn in_trait() -> impl Sized + use<'a> {}
-    |                      ^^^^^^^^^^^^^-------
---
To only update this specific test, also pass `--test-args impl-trait/precise-capturing/redundant.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/impl-trait/precise-capturing/redundant.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/impl-trait/precise-capturing/redundant" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2024"
stdout: none
--- stderr -------------------------------
error: all possible in-scope parameters are already captured, so `use<...>` syntax is redundant
##[error]  --> /checkout/tests/ui/impl-trait/precise-capturing/redundant.rs:5:19
   |
LL | fn hello<'a>() -> impl Sized + use<'a> {}
   |                   ^^^^^^^^^^----------
   |                             |
---
   |
LL | #![deny(impl_trait_redundant_captures)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: all possible in-scope parameters are already captured, so `use<...>` syntax is redundant
##[error]  --> /checkout/tests/ui/impl-trait/precise-capturing/redundant.rs:10:27
   |
LL |     fn inherent(&self) -> impl Sized + use<'_> {}
   |                           ^^^^^^^^^^----------
   |                                     |
   |                                     help: remove the `use<...>` syntax

error: all possible in-scope parameters are already captured, so `use<...>` syntax is redundant
##[error]  --> /checkout/tests/ui/impl-trait/precise-capturing/redundant.rs:15:22
   |
LL |     fn in_trait() -> impl Sized + use<'a, Self>;
   |                      ^^^^^^^^^^----------------
   |                                |
   |                                help: remove the `use<...>` syntax

error: all possible in-scope parameters are already captured, so `use<...>` syntax is redundant
##[error]  --> /checkout/tests/ui/impl-trait/precise-capturing/redundant.rs:19:22
   |
LL |     fn in_trait() -> impl Sized + use<'a> {}
   |                      ^^^^^^^^^^----------
   |                                |
```

</details>

--- a/compiler/rustc_lint/src/impl_trait_overcaptures.rs
+++ b/compiler/rustc_lint/src/impl_trait_overcaptures.rs
@@ -361,9 +361,9 @@ where
             // have no uncaptured args, then we should warn to the user that
             // it's redundant to capture all args explicitly.
             if new_capture_rules
-                && let Some((captured_args, capturing_span)) =
-                    opaque.bounds.iter().find_map(|bound| match *bound {
-                        hir::GenericBound::Use(a, s) => Some((a, s)),
+                && let Some((use_idx, captured_args, capturing_span)) =
+                    opaque.bounds.iter().enumerate().find_map(|(i, bound)| match *bound {
+                        hir::GenericBound::Use(a, s) => Some((i, a, s)),
                         _ => None,
                     })
             {
@@ -400,11 +400,25 @@ where
                     .iter()
                     .all(|(def_id, _)| explicitly_captured.contains(def_id))
                 {
+                    // Extend the removal span to include the `+` joiner adjacent
+                    // to `use<...>`, so applying the suggestion does not leave
+                    // behind a stray `+` that fails to parse.
+                    let suggestion_span = if let Some(next) = opaque.bounds.get(use_idx + 1) {
+                        capturing_span.with_hi(next.span().lo())
+                    } else if let Some(prev_idx) = use_idx.checked_sub(1) {
+                        let prev = opaque.bounds[prev_idx];
+                        capturing_span.with_lo(prev.span().hi())
+                    } else {
+                        // `impl use<...>` with no other bound is not valid
+                        // syntax, so this branch is unreachable in practice.
+                        capturing_span
+                    };
+
                     self.tcx.emit_node_span_lint(
                         IMPL_TRAIT_REDUNDANT_CAPTURES,
                         self.tcx.local_def_id_to_hir_id(opaque_def_id),
                         opaque_span,
-                        ImplTraitRedundantCapturesLint { capturing_span },
+                        ImplTraitRedundantCapturesLint { capturing_span: suggestion_span },
                     );
                 }
             }

--- a/tests/ui/impl-trait/precise-capturing/redundant-machine-applicable.fixed
+++ b/tests/ui/impl-trait/precise-capturing/redundant-machine-applicable.fixed
@@ -1,0 +1,28 @@
+//@ run-rustfix
+//@ rustfix-only-machine-applicable
+//@ edition: 2024
+
+// Verify that the suggestion produced by `impl_trait_redundant_captures`
+// removes the adjacent `+` joiner along with `use<...>`, instead of leaving
+// behind a stray `+` that fails to compile. Regression test for
+// https://github.com/rust-lang/rust/issues/143216.
+
+#![allow(unused)]
+#![deny(impl_trait_redundant_captures)]
+
+// `use<>` at the end of the bound list: the suggestion must remove the
+// preceding `+`.
+fn end_position() -> impl Sized {}
+//~^ ERROR all possible in-scope parameters are already captured
+
+// `use<>` at the start of the bound list: the suggestion must remove the
+// following `+`.
+fn start_position() -> impl Sized {}
+//~^ ERROR all possible in-scope parameters are already captured
+
+// `use<>` in the middle of the bound list: the suggestion must remove
+// exactly one `+`, keeping the other to join the remaining bounds.
+fn middle_position() -> impl Sized + Send {}
+//~^ ERROR all possible in-scope parameters are already captured
+
+fn main() {}

--- a/tests/ui/impl-trait/precise-capturing/redundant-machine-applicable.rs
+++ b/tests/ui/impl-trait/precise-capturing/redundant-machine-applicable.rs
@@ -1,0 +1,28 @@
+//@ run-rustfix
+//@ rustfix-only-machine-applicable
+//@ edition: 2024
+
+// Verify that the suggestion produced by `impl_trait_redundant_captures`
+// removes the adjacent `+` joiner along with `use<...>`, instead of leaving
+// behind a stray `+` that fails to compile. Regression test for
+// https://github.com/rust-lang/rust/issues/143216.
+
+#![allow(unused)]
+#![deny(impl_trait_redundant_captures)]
+
+// `use<>` at the end of the bound list: the suggestion must remove the
+// preceding `+`.
+fn end_position() -> impl Sized + use<> {}
+//~^ ERROR all possible in-scope parameters are already captured
+
+// `use<>` at the start of the bound list: the suggestion must remove the
+// following `+`.
+fn start_position() -> impl use<> + Sized {}
+//~^ ERROR all possible in-scope parameters are already captured
+
+// `use<>` in the middle of the bound list: the suggestion must remove
+// exactly one `+`, keeping the other to join the remaining bounds.
+fn middle_position() -> impl Sized + use<> + Send {}
+//~^ ERROR all possible in-scope parameters are already captured
+
+fn main() {}

--- a/tests/ui/impl-trait/precise-capturing/redundant-machine-applicable.stderr
+++ b/tests/ui/impl-trait/precise-capturing/redundant-machine-applicable.stderr
@@ -1,0 +1,32 @@
+error: all possible in-scope parameters are already captured, so `use<...>` syntax is redundant
+  --> $DIR/redundant-machine-applicable.rs:15:22
+   |
+LL | fn end_position() -> impl Sized + use<> {}
+   |                      ^^^^^^^^^^--------
+   |                                |
+   |                                help: remove the `use<...>` syntax
+   |
+note: the lint level is defined here
+  --> $DIR/redundant-machine-applicable.rs:11:9
+   |
+LL | #![deny(impl_trait_redundant_captures)]
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: all possible in-scope parameters are already captured, so `use<...>` syntax is redundant
+  --> $DIR/redundant-machine-applicable.rs:20:24
+   |
+LL | fn start_position() -> impl use<> + Sized {}
+   |                        ^^^^^--------^^^^^
+   |                             |
+   |                             help: remove the `use<...>` syntax
+
+error: all possible in-scope parameters are already captured, so `use<...>` syntax is redundant
+  --> $DIR/redundant-machine-applicable.rs:25:25
+   |
+LL | fn middle_position() -> impl Sized + use<> + Send {}
+   |                         ^^^^^^^^^^^^^--------^^^^
+   |                                      |
+   |                                      help: remove the `use<...>` syntax
+
+error: aborting due to 3 previous errors
+

--- a/tests/ui/impl-trait/precise-capturing/redundant.stderr
+++ b/tests/ui/impl-trait/precise-capturing/redundant.stderr
@@ -2,9 +2,9 @@ error: all possible in-scope parameters are already captured, so `use<...>` synt
   --> $DIR/redundant.rs:5:19
    |
 LL | fn hello<'a>() -> impl Sized + use<'a> {}
-   |                   ^^^^^^^^^^^^^-------
-   |                                |
-   |                                help: remove the `use<...>` syntax
+   |                   ^^^^^^^^^^----------
+   |                             |
+   |                             help: remove the `use<...>` syntax
    |
 note: the lint level is defined here
   --> $DIR/redundant.rs:3:9
@@ -16,25 +16,25 @@ error: all possible in-scope parameters are already captured, so `use<...>` synt
   --> $DIR/redundant.rs:10:27
    |
 LL |     fn inherent(&self) -> impl Sized + use<'_> {}
-   |                           ^^^^^^^^^^^^^-------
-   |                                        |
-   |                                        help: remove the `use<...>` syntax
+   |                           ^^^^^^^^^^----------
+   |                                     |
+   |                                     help: remove the `use<...>` syntax
 
 error: all possible in-scope parameters are already captured, so `use<...>` syntax is redundant
   --> $DIR/redundant.rs:15:22
    |
 LL |     fn in_trait() -> impl Sized + use<'a, Self>;
-   |                      ^^^^^^^^^^^^^-------------
-   |                                   |
-   |                                   help: remove the `use<...>` syntax
+   |                      ^^^^^^^^^^----------------
+   |                                |
+   |                                help: remove the `use<...>` syntax
 
 error: all possible in-scope parameters are already captured, so `use<...>` syntax is redundant
   --> $DIR/redundant.rs:19:22
    |
 LL |     fn in_trait() -> impl Sized + use<'a> {}
-   |                      ^^^^^^^^^^^^^-------
-   |                                   |
-   |                                   help: remove the `use<...>` syntax
+   |                      ^^^^^^^^^^----------
+   |                                |
+   |                                help: remove the `use<...>` syntax
 
 error: aborting due to 4 previous errors
 

